llvm/lib/Target/AMDGPU/SIISelLowering.cpp

```diff
@@ -16942,6 +16969,37 @@ bool SITargetLowering::checkForPhysRegDependency(
   return false;
 }
 
+/// Check if it is profitable to hoist instruction in then/else to if.
+/// Not profitable if I and it's user can form a FMA instruction
+/// because we prefer FMSUB/FMADD.
+bool SITargetLowering::isProfitableToHoist(Instruction *I) const {
+  if (!I->hasOneUse())
+    return true;
+
+  Instruction *User = I->user_back();
+  // TODO: Add more patterns that are not profitable to hoist
+  switch (I->getOpcode()) {
+  case Instruction::FMul: {
+    if (User->getOpcode() != Instruction::FSub &&
+        User->getOpcode() != Instruction::FAdd)
+      return true;
+
+    const TargetOptions &Options = getTargetMachine().Options;
+    const Function *F = I->getFunction();
+    const DataLayout &DL = F->getDataLayout();
+    Type *Ty = User->getOperand(0)->getType();
+
+    return !isOperationLegalOrCustom(ISD::FMA, getValueType(DL, Ty)) ||
```

Comment:
We probably don't need the legality check. It's legal for every type that matters. Maybe we care about the number of pieces in the wide vector cases?

Reply:
> Maybe we care about the number of pieces in the wide vector cases?
Could you help me with what cases we would actually care about this ?

Reply:
e.g. we want to lower `<N x half>` as N/2 pieces of x `<2 x half>` elements. At some N/2 it may be undesirable?

Reply:
I tried hunting for places where we currently do something like this, but was unable to figure it out. Will this depend on the wave size for instance ?

Reply:
No, this doesn't depend on the wavesize. Look at the lowering for any 4 x half vectors, you'll get 2 x 2 x half operations

Reply:
I think I understand what you imply. There's a performance / code size tradeoff to consider here for wide vectors. Is there a general rule of thumb when thinking about this ?

I could use [enableAggressiveFMAFusion](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/AMDGPU/SIISelLowering.cpp#L5640), but it currently returns a blanket `true` value. I'm assuming this means that if the appropriate flags are enabled, we encourage fusing of `fmul` and `fadd` as long as we have fast `fma`.

Reply:
You don't need to care about enableAggressiveFMAFusion. My point we really care about the type the FMA will ultimately be executed at, which is not the type that appears literally. Any scalar or vector with element type f16/f32/f64 we want regardless of whether the vector operation is directly legal

Reply:
Right, in that case I believe we already take care of this. I have removed the legality check and replicated logic that is defined [here](https://github.com/llvm/llvm-project/blob/0bd1c879966cfdf145b1f96292a2632628fab3fb/llvm/lib/Target/AMDGPU/SIISelLowering.cpp#L5673) to see if FMA is faster. For f64, we always prefer FMA fusing. There are some other checks we perform for f16/f32.

I wonder how to reason about the vector sizes in a comment you made earlier. When does fusing fmul and fadd with vector args become undesirable ?

Reply:
I don't think it ever does in this context